### PR TITLE
nclient4: real exponential backoff

### PR DESCRIPTION
**Pull request by hugelgupf**
Signed-off-by: Chris Koch <chrisko@google.com>

### CODE REVIEW

File: dhcpv4/nclient4/client.go [hunk 1]
```diff
@@ -233,10 +234,16 @@ func withBufferCap(n int) ClientOpt {
 
 // WithRetry configures the number of retransmissions to attempt.
 //
+// 0 retries means infinite retries.
+//
 // Default is 3.
 func WithRetry(r int) ClientOpt {
```
Comment (insomniacslk) [inline dhcpv4/nclient4/client.go]: what about requiring an uint here? The user will get a compile time error if a negative integer is passed, a bit annoying (more typing) but safer

File: dhcpv4/nclient4/client.go [hunk 2]
```diff
@@ -233,10 +234,16 @@ func withBufferCap(n int) ClientOpt {
 
 // WithRetry configures the number of retransmissions to attempt.
 //
+// 0 retries means infinite retries.
```
Comment (insomniacslk) [inline dhcpv4/nclient4/client.go]: I understand "zero retries" as "try just once, then stop" - same as above with "tries" vs. "retries". I'd rather keep using the concept of retries here

Comment (insomniacslk) [reply dhcpv4/nclient4/client.go]: which also means that 0 is a valid number of retries, and "loop forever" should be achieved differently (any negative number? Exactly -1 ? )

File: dhcpv4/nclient4/client.go [hunk 3]
```diff
@@ -67,7 +68,7 @@ type Client struct {
 	ifaceHWAddr net.HardwareAddr
 	conn        net.PacketConn
 	timeout     time.Duration
-	retry       int
+	tries       uint
```
[the remark this answers is not in the document]
Comment (hugelgupf) [reply dhcpv4/nclient4/client.go]: In our usage below, we seem to actually do number of tries, not retries. I changed the variable name to correct reflect our usage of it

File: dhcpv4/nclient4/client.go (hunk 1 above)
Comment (hugelgupf) [reply dhcpv4/nclient4/client.go]: I didn't want to disturb existing users, haha

File: dhcpv4/nclient4/client.go (hunk 2 above)
Comment (hugelgupf) [reply dhcpv4/nclient4/client.go]: went back to retries, but i think 0 is kind of a useless value, so <= 0 is infinite

**Comment by hugelgupf** [comment]
ptal

File: dhcpv4/nclient4/client.go (hunk 2 above)
Comment (insomniacslk) [reply dhcpv4/nclient4/client.go]: How I see the API as a user:
* if retries is 0, I expect one attempt
* if retries is 1, I expect two attempts
* and so on

The current code treats retries as attempts, so 0 is not meaningful, but I still think there is an inconsistency between naming and behaviour, which could confuse the users. Hope it makes sense, and open to discuss :)

Comment (insomniacslk) [reply dhcpv4/nclient4/client.go]: hm, why marked as resolved?

**Comment by insomniacslk** [comment]
there is also a small merge conflict with #295

**Comment by tfg13** [comment]
Now the comment says `Negative retries means infinite retries`, but the loop uses `c.retries <= 0` :P

**Comment by tfg13** [comment]
Or are we doing `retries = 1` means 1 total attempt now?

File: dhcpv4/nclient4/client.go [hunk 4]
```diff
@@ -386,15 +389,15 @@ func (c *Client) retryFn(fn func(timeout time.Duration) error) error {
...
 		switch err := fn(timeout); err {
 		case nil:
 			// Got it!
 			return nil
 
 		case errDeadlineExceeded:
-			// Double timeout, then retry.
-			timeout *= 2
+			// Exponential backoff.
+			timeout = time.Duration(rand.Intn(1<<i)-1) * c.timeout
```
Comment (tfg13) [inline dhcpv4/nclient4/client.go]: Isn't the timeout always negative for the first iteration (when `i = 0`? Is this on purpose?)

**Comment by hugelgupf** [comment]
Ryan's gonna take this over.

**Pull request closed without merging**